### PR TITLE
Group 7 - Solution for Tic Tac Toe (b9-pyp-g1)

### PR DESCRIPTION
**Pull request by janiceshiu**
Team members
- [Ammar Aziz](https://github.com/ammaraziz)
- [Janice Shiu](https://github.com/contrepoint/)

### CODE REVIEW

File: tic_tac_toe/game.py [hunk 1]
```diff
@@ -63,22 +95,60 @@ def _check_winning_combinations(board, player):
     Returns the player (winner) of any of the winning combinations is completed
     by given player, or None otherwise.
     """
-    pass
+    # positions = [
```
Comment (martinzugnoni) [inline tic_tac_toe/game.py]: When you are "presenting" your code, it's always bad to push commented or incomplete code :)

File: tic_tac_toe/game.py [hunk 2]
```diff
@@ -87,18 +157,54 @@ def move(game, player, position):
...
+        raise GameOver('"' + player + '" wins!')
+    elif _board_is_full(board):
+        raise GameOver('Game is tied!')
+    else: # can refactor into separate function later
+        if player == game['player1']:
+            game['next_turn'] = game['player2']
+        else:
+            game['next_turn'] = game['player1']
 
 def get_board_as_string(game):
```
Comment (martinzugnoni) [inline tic_tac_toe/game.py:180]: I like this way much better:

```
def get_board_as_string(game):
    """
    Returns a string representation of the game board in the current state.
    """
    board_template = """
{0}  |  {1}  |  {2}
--------------
{3}  |  {4}  |  {5}
--------------
{6}  |  {7}  |  {8}
"""
    board = game['board']
    return board_template.format(*(board[0] + board[1] + board[2]))
```

**Review by martinzugnoni — APPROVED** [review]
Great work guys! 🎉 congrats.

File: tic_tac_toe/game.py (hunk 1 above)
Comment (contrepoint) [inline tic_tac_toe/game.py]: ahhhh you're too fast T___T I'm currently modifying it.

Comment (janiceshiu) [reply tic_tac_toe/game.py]: ahhhh you're too fast T___T I'm currently modifying it. we pushed first to make sure it worked with all Python versions

edit: all removed ^^

File: tic_tac_toe/game.py (hunk 2 above)
Comment (contrepoint) [inline tic_tac_toe/game.py]: modified - thanks for your detailed explanation in the `mentors-help` channel!

Comment (janiceshiu) [reply tic_tac_toe/game.py:180]: ooh yeah - much neater way of doing it. (I knew our way was messy but we weren't too sure how to make it nicer.) modified - thanks for your detailed explanation in the mentors-help channel!

File: tic_tac_toe/game.py [hunk 3]
```diff
@@ -33,7 +50,12 @@ def _board_is_full(board):
 
     :param board: Game board.
     """
-    pass
+    for row in range(3):
+        for column in range(3):
+            if _position_is_empty_in_board((row, column), board):
+                return False
+
+    return True
```
Comment (janiceshiu) [inline tic_tac_toe/game.py:59]: another option that passes the tests is

```
positions = [
    (0,0), (0,1), (0,2),
    (1,0), (1,1), (1,2),
    (2,0), (2,1), (2,2),
]

return '-' in positions
```

However, the loop in the first scenario breaks once it finds an already-occupied position on the board. Thus, it might be faster than this version. Must figure out how to benchmark both ways.

File: tic_tac_toe/game.py [hunk 4]
```diff
@@ -63,22 +92,53 @@ def _check_winning_combinations(board, player):
...
+        ((0, 1), (1, 1), (2, 1)),
+        ((0, 2), (1, 2), (2, 2))
+    )
+
+    diagonals = (
+        ((0, 0), (1, 1), (2, 2)),
+        ((0, 2), (1, 1), (2, 0))
+    )
+
+    winning_combinations = rows + columns + diagonals
```
Comment (janiceshiu) [inline tic_tac_toe/game.py:114]: if there's some code to generate all the straight lines in the board, it would be awesome. we haven't figured out how so this is hard-coded, sadly.

**Review by janiceshiu — COMMENTED** [review]
commenting on some of the things I noticed with our (me + @ammaraziz). Thanks in advance for any feedback.

File: tic_tac_toe/game.py (hunk 4 above)
Comment (jangeador) [reply tic_tac_toe/game.py:114]: we went with this:

``` python
for i in range(3):
        # check rows
        if _is_winning_combination(board, ((0, i), (1, i), (2, i)), player):
            return player
        # check columns
        if _is_winning_combination(board, ((i, 0), (i, 1), (i, 2)), player):
            return player
    # check diagonals
    if _is_winning_combination(board, ((0, 0), (1, 1), (2, 2)), player):
        return player
    if _is_winning_combination(board, ((2, 0), (1, 1), (0, 2)), player):
        return player
    return None
```

File: tic_tac_toe/game.py [hunk 5]
```diff
@@ -9,7 +18,10 @@ def _position_is_empty_in_board(position, board):
 
     Returns True if given position is empty, False otherwise.
     """
-    pass
+    row = position[0]
+    column = position[1]
```
Comment (jangeador) [inline tic_tac_toe/game.py:22]: This is an good place to use tuple unpacking:

``` python
row, column = position
```

File: tic_tac_toe/game.py [hunk 6]
```diff
@@ -47,7 +70,15 @@ def _is_winning_combination(board, combination, player):
     Returns True of all three positions in the combination belongs to given
     player, False otherwise.
     """
-    pass
+    for position in combination:
+        row = position[0]
+        column = position[1]
```
Comment (jangeador) [inline tic_tac_toe/game.py:75]: same as the comment on line 22

File: tic_tac_toe/game.py [hunk 7]
```diff
@@ -63,22 +94,53 @@ def _check_winning_combinations(board, player):
...
+    return None
 
 
 # public interface
 def start_new_game(player1, player2):
     """
     Creates and returns a new game configuration.
     """
-    pass
+    board = [['-' for columns in range(3)] for rows in range(3)]
```
Comment (jangeador) [inline tic_tac_toe/game.py:128]: Excellent application of nested list comprehension. I like this.

File: tic_tac_toe/game.py [hunk 8]
```diff
@@ -87,18 +149,52 @@ def move(game, player, position):
...
     """
-    pass
+    board = game['board']
+    current_player = get_next_turn(game)
+
+    if _board_is_full(board) or game['winner'] != None:
+        raise InvalidMovement('Game is over.')
+
+    if current_player != player:
+        raise InvalidMovement('"' + current_player + '" moves next.')
```
Comment (jangeador) [inline tic_tac_toe/game.py:159]: Be careful with this way of concatenating string as it is not optimal. If the player choses numbers instead of 'x' and 'o' this will break:

``` python
>>> current_player = 2
>>> '"' + current_player + '" moves next.'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't convert 'int' object to str implicitly
```

The `string.format()` method would be better for any string concatenation as it casts everything to string implicitly.

File: tic_tac_toe/game.py [hunk 9]
```diff
@@ -87,18 +149,52 @@ def move(game, player, position):
...
+
+    winner = _check_winning_combinations(board, player)
+
+    if winner:
+        game['winner'] = player
+        raise GameOver('"' + player + '" wins!')
+    elif _board_is_full(board):
+        raise GameOver('Game is tied!')
+    else:
+        _switch_turns(game, player)
```
Comment (jangeador) [inline tic_tac_toe/game.py:177]: I like how you put _switch_turns into a helper method.

File: tic_tac_toe/game.py [hunk 10]
```diff
@@ -87,18 +149,52 @@ def move(game, player, position):
...
+    board_template = """
+{0}  |  {1}  |  {2}
+--------------
+{3}  |  {4}  |  {5}
+--------------
+{6}  |  {7}  |  {8}
+"""
+    board = game['board']
+    
+    return board_template.format(*(board[0] + board[1] + board[2]))
```
Comment (jangeador) [inline tic_tac_toe/game.py:193]: This is just cool on so many levels. Excellent ~~ab~~use of `string.format()`. I am putting this into a gist for future use.

**Review by jangeador — COMMENTED** [review]
Excellent solution. Your `move` method is very clean. Great job Group 7.

File: tic_tac_toe/game.py (hunk 8 above)
Comment (janiceshiu) [reply tic_tac_toe/game.py:159]: definitely. I come from Ruby where you can just interpolate it without `.format` (eg: `a = 'text'` and `b = "a is #{a}"`) so I'm still getting used to Python's way of doing things. Meant to reformat but Amar and I were too tired last night.

which also reminds me I should set a check that prevents a player from using `'-'` as their marker (given how we test for winning combos, etc). edge cases, edge cases.

File: tic_tac_toe/game.py (hunk 9 above)
Comment (janiceshiu) [reply tic_tac_toe/game.py:177]: thanks! it was an `if/else` block at first but I refactored it after we got all the tests to pass on travis. I prefer abstracting away stuff wherever possible because it's easier to modify later on.

File: tic_tac_toe/game.py (hunk 10 above)
Comment (janiceshiu) [reply tic_tac_toe/game.py:193]: all credit to @martinzugnoni for that. (he also explained `.format(*(board[0] + board[1] + board[2])` in `mentors-help`)

~~i see your strikeout XD~~

File: tic_tac_toe/game.py (hunk 5 above)
Comment (janiceshiu) [reply tic_tac_toe/game.py:22]: thanks! looks familiar - must've seen it in the readings somewhere. I didn't like how we repeated `position` so this is useful.

File: tic_tac_toe/game.py (hunk 4 above)
Comment (janiceshiu) [reply tic_tac_toe/game.py:114]: _nice!_ we conjectured we could do a for loop for rows and columns but were too tired to implement it. thanks!

@ammaraziz fyi

File: tic_tac_toe/game.py (hunk 8 above)
Comment (jangeador) [reply tic_tac_toe/game.py:159]: Yes, if someone decides to use `'-'` as their token my solution will shatter in a million pieces as well.

File: tic_tac_toe/game.py (hunk 10 above)
Comment (jsymons) [reply tic_tac_toe/game.py:193]: Very nice. Had thought of this but couldn't think of a way to put the board values in without making the code look like a bit of a clusterf*@#.

**Comment by jsymons** [comment]
Great job!
